### DISCUSSION TITLE
How to read buildId on the server side?

### DISCUSSION THREAD

**Discussion by magiskboy** [vercel/next.js#36536] [Help]
Currently, I want to read buildId on the server side without reading BUILD_ID file? Does anyone have idea for this?

**Comment by icyJoseph** [comment]
I am not sure if this would work when deploying, but if we do a few adjustments:

1. `next.config.js`: normalize the build id to be the hash of the last commit

```js
const { execSync } = require("child_process");

const lastCommitCommand = "git rev-parse HEAD";

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  generateBuildId() {
    return execSync(lastCommitCommand).toString().trim();
  },
};

module.exports = nextConfig;
```

2. From `getServerSideProps` side code, with access to `process.cwd`:

```tsx
import type { GetServerSidePropsContext, NextPage } from "next";

import { execSync } from "child_process";

const lastCommitCommand = "git rev-parse HEAD";

const Home: NextPage = () => {
  return <div>Home</div>;
};

export const getServerSideProps = (ctx: GetServerSidePropsContext) => {
  const buildId = execSync(lastCommitCommand, { cwd: process.cwd() })
    .toString()
    .trim();

  console.log({ buildId });
  return { props: {} };
};

export default Home;
```

3. In API routes:

```ts
// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import type { NextApiRequest, NextApiResponse } from "next";

import { execSync } from "child_process";

const lastCommitCommand = "git rev-parse HEAD";

type Data = {
  name: string;
};

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {
  const buildId = execSync(lastCommitCommand, { cwd: process.cwd() })
    .toString()
    .trim();

  console.log({ buildId });
  res.status(200).json({ name: "John Doe" });
}
```

Could you give it a go? If you are using Vercel specially, it'd nice to see if it works when deployed. Locally it should definitely work, I think.

**Reply by maxymapp** [reply]
@icyJoseph 
just out of curiosity, in Vercel, is reading the last commit ID a reliable way to confirm the build for this commit compiled successfully? 
For example, I don't use Vercel (due to the mandatory self-hosted setup at my company) and have a script
```
...
 git pull origin main;

 yarn build;
...
```
that may `git pull` and yet the build command may not be carried out successfully.

**Comment by vadymshymko** [comment]
Hi there. +1 for @icyJoseph answer. And one more option from me:
1. `next.config.js `
```
module.exports = {
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
    const customPlugins = [
      new webpack.DefinePlugin({
        'process.env.YOUR_BUILD_ID_VAR_NAME': JSON.stringify(buildId),
      })
    ];
    config.plugins.push(...customPlugins);
    // Important: return the modified config
    return config
  },
}

// or, if you don't want to make the buildId value public, maybe you can try something like this:
module.exports = {
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
    process.env.YOUR_BUILD_ID_VAR_NAME = JSON.stringify(buildId);
    // Important: return the modified config
    return config
  },
}

```


2. Somewhere in the code:
`console.log(process.env.YOUR_BUILD_ID_VAR_NAME) \\ print actual buildId`

**Reply by icyJoseph** [reply]
Yeah this might be better because it bundles the build id as an env var, though if you deploy to Vercel, you can access VERCEL_GIT_COMMIT_SHA, if you activate it, https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables and use that as build id I guess.
